src/consul.rs

```diff
@@ -0,0 +1,399 @@
+    pub hostname: String,
+    #[serde(rename = "type")]
+    pub type_: DnsType,
+    pub ttl: Option<i32>,
+    pub value: String,
+}
+
+#[derive(Debug, Eq, PartialEq, Hash, Serialize, Deserialize)]
+struct ConsulLock {
+    pub locked_at: SystemTime,
+}
+
+#[derive(Debug, Deserialize)]
+struct ConsulKVResponse {
+    #[serde(rename = "Value")]
+    value: Option<String>,
+    #[serde(rename = "Key")]
+    key: String,
+    #[serde(rename = "Session")]
+    session: Option<String>,
+}
+
+#[derive(Clone)]
+pub struct ConsulClient {
+    pub http_client: reqwest::Client,
+    pub kv_api_base_url: Url,
+    pub catalog_api_base_url: Url,
+    pub session_api_base_url: Url,
+    pub datacenter: Option<String>,
+    pub kv_dns_records: HashMap<String, DnsRecord>,
+}
+
+impl ConsulClient {
+    pub fn new(consul_address: Url, consul_datacenter: Option<String>) -> Result<ConsulClient> {
+        let kv_api_base_url = consul_address.join("v1/")?.join("kv/")?;
+        let catalog_api_base_url = consul_address.join("v1/")?.join("catalog/")?;
+        let session_api_base_url = consul_address.join("v1/")?.join("session/")?;
+        let client = reqwest::Client::builder().build()?;
+        Ok(ConsulClient {
+            http_client: client,
+            kv_api_base_url,
+            catalog_api_base_url,
+            session_api_base_url,
+            datacenter: consul_datacenter,
+            kv_dns_records: HashMap::new(),
+        })
+    }
+
+    /// Create a new session in Consul
+    /// This session is used to acquire a lock
+    pub async fn create_session(&self) -> Result<Uuid, anyhow::Error> {
+        let session_request = CreateSessionRequest {
+            name: "external-dns",
+            behavior: "release",
+            ttl: "30s",
+            lock_delay: "30s",
+        };
+
+        let session_url = self.session_api_base_url.join("create")?;
+
+        let mut req = self.http_client.put(session_url).json(&session_request);
+
+        // Set dc if it is provided in the config
+        if let Some(dc) = &self.datacenter {
+            req = req.query(&[("dc", dc)]);
+        }
+
+        let resp = req.send().await?.error_for_status()?;
+        let session_response: CreateSessionResponse = resp.json().await?;
+        Ok(session_response.id)
+    }
+
+    /// Renew the Consul session
+    pub async fn renew_session(&self, session_id: Uuid) -> Result<(), anyhow::Error> {
+        let session_url = self
+            .session_api_base_url
+            .join(&format!("renew/{}", session_id))?;
+
+        let mut req = self.http_client.put(session_url);
+
+        // Set dc if it is provided in the config
+        if let Some(dc) = &self.datacenter {
+            req = req.query(&[("dc", dc)]);
+        }
+
+        req.send().await?.error_for_status()?;
+        println!("Renewed Consul session: {}", session_id);
+        Ok(())
+    }
+
+    /// Acquire a lock
+    ///
+    /// Times out after a while and returns an error if it does.
```

Comment:
> What exactly are the rules for the timeout.

I'm not sure why the comment says that, because it should not ever time out.

> Is the lock renewed at some point?

A lock is never renewed, it's valid until it's either released or the session that took it is invalidated.

> What happens if operations take longer than the acquired lock time, would we continue operation even though we no longer have a lock?

There is no such thing as the "longer than the acquired lock time", but the session's lock delay feature is used to prevent anyone else from acquiring the lock until X amount of time after the session was invalidated, and as long as we renew the session more frequently than that we'll find out if our session was invalidated before anything else can take the lock and can therefore trigger a graceful shutdown.

Reply:
I removed the timeout comment